plenum/server/catchup/utils.py

```diff
@@ -0,0 +1,139 @@
+        self._node = node  # TODO: Come up with a better name?
+
+    def __repr__(self):
+        return self._name
+
+    def get_ledger_status(self) -> LedgerStatus:
+        # TODO: Delete when INDY-1946 gets implemented
+        three_pc_key = self._node.three_phase_key_for_txn_seq_no(self._ledger_id, self._ledger.size)
+        view_no, pp_seq_no = three_pc_key if three_pc_key else (None, None)
+
+        return LedgerStatus(self._ledger_id,
+                            self._ledger.size,
+                            view_no,
+                            pp_seq_no,
+                            self._ledger.root_hash,
+                            CURRENT_PROTOCOL_VERSION)
+
+    def get_consistency_proof(self, seq_no_start: int, seq_no_end: int) -> Optional[ConsistencyProof]:
+        if seq_no_end < seq_no_start:
+            logger.error("{} cannot build consistency proof since end {} is less than start {}".
+                         format(self, seq_no_end, seq_no_start))
+            return
+
+        if seq_no_start > self._ledger.size:
+            logger.error("{} cannot build consistency proof from {} since its ledger size is {}"
+                         .format(self, seq_no_start, self._ledger.size))
+            return
+
+        if seq_no_end > self._ledger.size:
+            logger.error("{} cannot build consistency proof till {} since its ledger size is {}"
+                         .format(self, seq_no_end, self._ledger.size))
+            return
+
+        if seq_no_start == 0:
+            # Consistency proof for an empty tree cannot exist. Using the root
+            # hash now so that the node which is behind can verify that
+            # TODO: Make this an empty list
+            old_root = self._ledger.tree.root_hash
+            proof = [old_root, ]
+        else:
+            proof = self._ledger.tree.consistency_proof(seq_no_start, seq_no_end)
+            old_root = self._ledger.tree.merkle_tree_hash(0, seq_no_start)
+
+        new_root = self._ledger.tree.merkle_tree_hash(0, seq_no_end)
+
+        old_root = Ledger.hashToStr(old_root)
+        new_root = Ledger.hashToStr(new_root)
+        proof = [Ledger.hashToStr(p) for p in proof]
+
+        # TODO: Delete when INDY-1946 gets implemented
+        three_pc_key = self._node.three_phase_key_for_txn_seq_no(self._ledger_id, seq_no_end)
+        logger.info('{} found 3 phase key {} for seqNo {}'.format(self, three_pc_key, seq_no_end))
+        view_no, pp_seq_no = three_pc_key if three_pc_key else (0, 0)
+
+        return ConsistencyProof(self._ledger_id,
+                                seq_no_start,
+                                seq_no_end,
+                                view_no,
+                                pp_seq_no,
+                                old_root,
+                                new_root,
+                                proof)
+
+    def process_ledger_status(self, status: LedgerStatus, frm: str) -> Optional[ConsistencyProof]:
```

Comment:
I'm not sure we should have any `process_X` methods in data provider, since process messages represent a business logic which needs to be in LedgerCatchupService.
I was thinking that `CatchupLedgerDataProvider` is a more or less small helper class just to build catsup-related messages taking information from the ledger.

Reply:
Moving to services